src/commonControls/questionnaireControl/QuestionnaireControl.ts

```diff
@@ -0,0 +1,1578 @@
+            (input.request as IntentRequest).intent,
+        );
+
+        // Accept the answer if it is a match to one of the listed choices, otherwise
+        // report error and ignore.  Note: being a match is a stronger condition than
+        // simply being an ER_MATCH as the answer slot type might have more values than
+        // answers for a particular questionnaire.
+
+        if (content.choices.some((choice) => choice.id === valueStr)) {
+            const choiceIndex = this.getChoiceIndexById(content, valueStr);
+            assert(choiceIndex !== undefined);
+            const questionIndex = this.getQuestionIndexById(content, this.state.focusQuestionId!);
+
+            this.updateAnswer(question.id, valueStr, input, resultBuilder);
+
+            resultBuilder.addAct(
+                new QuestionAnsweredAct(this, {
+                    questionId: question.id,
+                    choiceId: valueStr,
+                    userAnsweredWithExplicitValue: true,
+                    userMentionedQuestion: false,
+                    renderedChoice: content.choices[choiceIndex].prompt,
+                    renderedQuestionShortForm: this.evaluatePromptShortForm(
+                        content.questions[questionIndex].promptShortForm,
+                        input,
+                    ),
+                }),
+            );
+        } else {
+            throw new Error('todo');
+            //resultBuilder.addAct(new InvalidAnswerAct(this, {}));
+        }
+        this.clearCompletionFlag(); // user interacted with the and so the questionnaire should stick around
+        return;
+    }
+
+    // // handle the special case that the answer is captured by `feedback` item.
+    private isFeedbackAnswerToSpecificQuestion(input: ControlInput): boolean {
+        try {
+            okIf(InputUtil.isIntent(input, GeneralControlIntent.name));
+            const { feedback, action, target } = unpackGeneralControlIntent(
+                (input.request as IntentRequest).intent,
+            );
+            const content = this.getQuestionnaireContent(input);
+
+            okIf(InputUtil.targetIsDefined(target));
+            const targetedQuestion = content.questions.find((q) => q.targets.includes(target!));
+            okIf(targetedQuestion !== undefined);
+
+            okIf(InputUtil.feedbackIsDefined(feedback));
+            okIf(content.choices.some((choice) => choice.id === feedback));
+            return true;
+        } catch (e) {
+            verifyErrorIsGuardFailure(e);
+            return false;
+        }
+    }
+
+    private handleFeedbackAnswerToSpecificQuestion(input: ControlInput, resultBuilder: ControlResultBuilder) {
+        const content = this.getQuestionnaireContent(input);
+
+        const { feedback, action, target } = unpackGeneralControlIntent(
+            (input.request as IntentRequest).intent,
+        );
+
+        assert(feedback !== undefined);
```

Comment:
if these asserts fails, don't they throw error exceptions and fail?

Reply:
yes, but that is what we tested in canHandle.  rather than using ! which would lead to a nullReferenceException we can now assert something more explicitly.